### PR TITLE
Enable RichProgressBar by default if Rich is available

### LINKED ISSUE #9580: Enable Rich Progress Bar/Model Summary if Rich is available
## 🚀 Feature

Currently we have the Rich Progress Bar/Summary in Lightning. However, we only enable it if the user passes the Rich Progress Bar Callback to the Trainer manually.

I see a few options, based on what we'd like to go for in the future:

1. Enable the Rich Progress Bar/Model Summary if Rich is available (@ananthsub mentioned that we might need to check for Rich 10+ https://github.com/PyTorchLightning/pytorch-lightning/pull/9546#discussion_r709783674)
2. Put the Rich Progress Bar/Summary in [Bolts](https://github.com/PyTorchLightning/lightning-bolts). This means we can only enable the Rich Progress Bar if Bolts is installed (and a version that contains the progress bar)
3. Keep as is

Personally I'm a big fan of 1 as it gives users a nice new progress bar/model summary to look at, which would be nice for a new release! All the user needs to do is `pip install rich -U`.

If you have other alternatives please put below! Also please let me know which options you'd like

cc @borda @justusschock @kaushikb11 @rohitgr7

### PR DESCRIPTION
**Pull request by SeanNaren**
## What does this PR do?

Closes #9580. By default if Rich (10.0.2 or later) is available, we use the Rich Progress Bar! 

Related to #9616, we throw a warning if the user sets `progress_bar_refresh_rate` from the Trainer and default to TQDM (the `ProgressBar`). This is because the meaning of refresh rate is different for Rich/TQDM, and are not related. Since in #9616 the argument is being deprecated, I felt it made sense to ask the users to pass the `RichProgressBar` to callbacks like `callbacks=RichProgressBar(refresh_rate_per_second=20)`.

cc @daniellepintz 

## Before submitting

- [x] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [x] Did you make sure to **update the documentation** with your changes? (if necessary)
- [x] Did you write any **new necessary tests**? (not for typos and docs)
- [x] Did you verify new and **existing tests pass** locally with your changes?
- [x] Did you list all the **breaking changes** introduced by this pull request?
- [x] Did you **update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)**? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

- [x] Is this pull request ready for review? (if not, please submit in draft mode)
- [x] Check that all items from **Before submitting** are resolved
- [x] Make sure the title is self-explanatory and the description concisely explains the PR
- [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?

Make sure you had fun coding 🙃

### CODE REVIEW

**Review by tchaton — APPROVED** [review]
LGMT !

**Review by ethanwharris — APPROVED** [review]
LGTM :smiley:

File: pytorch_lightning/trainer/connectors/callback_connector.py
Comment (rohitgr7) [inline pytorch_lightning/trainer/connectors/callback_connector.py:138]: is it `RichProgressBar(refresh_rate_per_second)` ??
```suggestion
                "please pass ``callbacks=RichProgressBar(refresh_rate_per_second=4)``. "
```

File: pytorch_lightning/trainer/connectors/callback_connector.py
Comment (daniellepintz) [inline pytorch_lightning/trainer/connectors/callback_connector.py:158]: > if use_rich_progress_bar

I believe this will never be True, because of line 141

File: pytorch_lightning/trainer/connectors/callback_connector.py
Comment (daniellepintz) [reply pytorch_lightning/trainer/connectors/callback_connector.py:138]: I don't think so:
https://github.com/PyTorchLightning/pytorch-lightning/blob/0ffc848dac4ec06839df03ae8d8c4363852651e9/pytorch_lightning/callbacks/progress/rich_progress.py#L138

File: tests/callbacks/test_rich_progress_bar.py
Comment (daniellepintz) [inline tests/callbacks/test_rich_progress_bar.py:48]:
```suggestion
def test_rich_progress_bar_refresh_rate(tmpdir):
```

File: pytorch_lightning/trainer/connectors/callback_connector.py
Comment (daniellepintz) [inline pytorch_lightning/trainer/connectors/callback_connector.py:143]:
```suggestion
        if _RICH_AVAILABLE:
            if refresh_rate is None:
                return RichProgressBar()
             else:
                 rank_zero_warn(
                    "``RichProgressBar`` does not support setting the refresh rate via the Trainer. "
                    "If you'd like to change the refresh rate and continue using the ``RichProgressBar``, "
                    "please pass ``callbacks=RichProgressBar(refresh_rate=4)``. "
                    "Setting to the ``TQDMProgressBar``."
                )
                  
        
        if os.getenv("COLAB_GPU") and refresh_rate is None:
            # smaller refresh rate on colab causes crashes for TQDM, choose a higher value
```

File: pytorch_lightning/trainer/connectors/callback_connector.py
Comment (daniellepintz) [inline pytorch_lightning/trainer/connectors/callback_connector.py:133]: Can we move this code to the top of the function?
```
progress_bars = [c for c in self.trainer.callbacks if isinstance(c, ProgressBarBase)]
        if len(progress_bars) > 1:
            raise MisconfigurationException(
                "You added multiple progress bar callbacks to the Trainer, but currently only one"
                " progress bar is supported."
            )
        if len(progress_bars) == 1:
            progress_bar_callback = progress_bars[0]
```

This function is now getting pretty complex, I think we could make it more understandable by doing three things in this order:
1. If the user has already passed a PBarCallback in "callbacks", just return that immediately
2. Then check if Rich is available and if it is and no RR is set, immediately return a new RichProgressBar()
3. Finally if we get to this point create a TQDMPBar with RR logic as before

File: pytorch_lightning/trainer/connectors/callback_connector.py
Comment (Borda) [reply pytorch_lightning/trainer/connectors/callback_connector.py:138]: per second can be quite annoying in case you run in Jupyter and it accidentally prints each update on a new line, so you may end up with hundreds of lines without any real change...

Comment (SeanNaren) [reply pytorch_lightning/trainer/connectors/callback_connector.py:138]: @Borda this isn't an issue for Rich on colab notebooks, this was only an issue with TQDM

File: pytorch_lightning/trainer/connectors/callback_connector.py
Comment (awaelchli) [inline pytorch_lightning/trainer/connectors/callback_connector.py:136]: we only use single backticks in console outputs

**Review by awaelchli — APPROVED** [review]

**Comment by SeanNaren** [comment]
Seems due to ddp spawn, the progress bar needs to be pickleable. This has introduced a bit of additional logic, and will need to debug to ensure it works well with ddp spawn!

File: pytorch_lightning/trainer/connectors/callback_connector.py
Comment (daniellepintz) [inline pytorch_lightning/trainer/connectors/callback_connector.py:149]:
```suggestion
        if refresh_rate != 0
```
For some reason I prefer `if` over `elif` here, not sure why

Comment (SeanNaren) [reply pytorch_lightning/trainer/connectors/callback_connector.py:149]: Yep fully agree, nice cleanup :)

File: pytorch_lightning/trainer/connectors/callback_connector.py
Comment (daniellepintz) [inline pytorch_lightning/trainer/connectors/callback_connector.py:140]: Can we add some comments to this function delineating the 3 steps I mentioned earlier, for readability?
something like: 
```
# if progress bar callback already exists return it
# if rich is available and no RR set return new RPB
# else return TQDM PB
```

File: pytorch_lightning/trainer/connectors/callback_connector.py
Comment (daniellepintz) [inline pytorch_lightning/trainer/connectors/callback_connector.py:149]: can we also just add a quick comment here about what happens when RR = 0?

**Comment by SeanNaren** [comment]
Seeing IPU failures so holding off for now!

File: pytorch_lightning/trainer/connectors/callback_connector.py
Comment (daniellepintz) [inline pytorch_lightning/trainer/connectors/callback_connector.py:143]:
```suggestion
        # if Rich is available refresh_rate is None return Rich ProgressBar
        # else return TQDM ProgressBar
```

File: pytorch_lightning/trainer/connectors/callback_connector.py
Comment (daniellepintz) [inline pytorch_lightning/trainer/connectors/callback_connector.py:154]:
```suggestion
        # if Rich is available and refresh_rate is None return new RichProgressBar
            if _RICH_AVAILABLE:
```

File: pytorch_lightning/trainer/connectors/callback_connector.py
Comment (daniellepintz) [inline pytorch_lightning/trainer/connectors/callback_connector.py:165]:
```suggestion
                    )
           # else return new TQDMProgressBar
```

**Comment by daniellepintz** [comment]
Sorry for all the back and forth regarding the comments, just added some more suggestions but feel free to ignore :)

**Comment by SeanNaren** [comment]
Before this PR is merged there are two more things to address:

- When using multiple nodes, only one of the main processes will include a progress bar. I don't think this is preferred and can be fixed with changing the rank zero check to local rank zero
- The tests within `test_progress_bar.py` are failing as the default progress bar has been changed. I'll add rich as a parameter to the tests that can be agnostic to progress bar, and copy tests that need to be redone to the rich progress bar tests.

File: tests/callbacks/test_rich_progress_bar.py
Comment (Borda) [inline tests/callbacks/test_rich_progress_bar.py:63]: Let's put under `with` only the line which supposes to raise warning not all... this makes it a bit foggy

File: pytorch_lightning/callbacks/progress/rich_progress.py
Comment (kaushikb11) [inline pytorch_lightning/callbacks/progress/rich_progress.py:197]: Why remove it from the `setup` hook?

File: pytorch_lightning/utilities/imports.py
Comment (kaushikb11) [inline pytorch_lightning/utilities/imports.py:93]: Maybe I might have missed the discussion on how we decided on `10.2.2`? What `rich` library change/feature in `v10.2.2` prompted us to do it?

File: pytorch_lightning/callbacks/progress/rich_progress.py
Comment (SeanNaren) [reply pytorch_lightning/callbacks/progress/rich_progress.py:197]: This is for spawn support, putting this call in setup meant the progress bar was being created before the processes are being spawned. The progress bar is not pickleable, hence has to be made after the spawn of processes (this also follows inline of what happens for TQDM)

File: pytorch_lightning/utilities/imports.py
Comment (SeanNaren) [reply pytorch_lightning/utilities/imports.py:93]: @ananthsub mentioned that with earlier versions of Rich this lead to crashes! This might be because rich has been installed a very long time ago and not updated. I may also be wrong but I noticed that earlier versions of rich from 10.2.2 had their own issues.

**Review by kaushikb11 — APPROVED** [review]
LGTM! Rich, it is.

**Comment by SeanNaren** [comment]
This PR has surfaced a lot of issues and as a result a lot of improvements have been made. They'll be addressed in separate PRs, before this being merged!

**Comment by kaushikb11** [comment]
Removed it from Draft as I want all tests to run to debug issues.

File: pytorch_lightning/trainer/connectors/callback_connector.py
Comment (daniellepintz) [inline pytorch_lightning/trainer/connectors/callback_connector.py:243]: should be indented in?

**Comment by kaushikb11** [comment]
Closing it in favor of #10912

**Pull request closed without merging**
